x/programs/rust/sdk_macros/src/lib.rs

```diff
@@ -46,19 +46,41 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
         }
         panic!("Unsupported function parameter format.");
     });
-
     let (param_names, param_types): (Vec<_>, Vec<_>) = full_params.unzip();
+
+    let program_param = &param_names[0];
+    // convert to a string
+    let program_name = name.to_string();
     let converted_params = param_names
         .iter()
         .map(|param_name| convert_param(param_name));
 
     // Extract the original function's return type. This must be a WASM supported type.
     let return_type = &input.sig.output;
+
+    let is_reentrant = match attr.to_string().as_str() {
+        "reentrant" => true,
+        "" => false,
+        _ => panic!(
+            "Invalid attribut `{other}` for `public` macro",
+            other = attr
+        ),
+    };
+    let rentrancy_check = if !is_reentrant {
+        quote! {
+            // enter_program will error in the host if the program is already entered
+            wasmlanche_sdk::host::enter_program(#program_param, #program_name).expect("error calling enter_program host function");
+        }
+    } else {
+        quote! {}
+    };
```

Comment:
I think this is fine the way it is now, but this whole thing might be a lot simpler if we do the locking in the guest and only track the active-call count in the host. We can even pass that in directly as an argument instead of needing to allocate in the guest. I think that would really simplify the host code.